text/0000-incremental-compilation.md

```diff
@@ -0,0 +1,648 @@
+
+## Interaction with optimization
+
+It is important to clarify, though, that this design does not attempt
+to enable full optimizing for incremental compilation; indeed the two
+are somewhat at odds with one another, as full optimization may
+perform inlining and inter-function analysis, which can cause small
+edits in one function to affect the generated code of another. This
+situation is further exacerbated by the fact that LLVM does not
+provide any way to track these sorts of dependencies (e.g., one cannot
+even determine what inlining took place, though @dotdash suggested a
+clever trick of using llvm lifetime hints). Strategies for handling
+this are discussed in the [Optimization section](#optimization) below.
+
+# Detailed design
+
+We begin with a high-level execution plan, followed by sections that
+explore aspects of the plan in more detail. The high-level summary
+includes links to each of the other sections.
+
+## High-level execution plan
+
+Regardless of whether it is invoked in incremental compilation mode or
+not, the compiler will always parse and macro expand the entire crate,
+resulting in a HIR tree. Once we have a complete HIR tree, and if we
+are invoked in incremental compilation mode, the compiler will then
+try to determine which parts of the crate have changed since the last
+execution. For each item, we compute a [(mostly) stable id](#defid)
+based primarily on the item's name and containing module. We then
+compute a hash of its contents and compare that hash against the hash
+that the item had in the compilation (if any).
+
+Once we know which items have changed, we consult a
+[dependency graph](#depgraph) to tell us which artifacts are still
+usable. These artifacts can take the form of serializing MIR graphs,
+LLVM IR, compiled object code, and so forth. The dependency graph
+tells us which bits of AST contributed to each artifact. It is
+constructed by dynamically monitoring what the compiler accesses
+during execution.
+
+Finally, we can begin execution. The compiler is currently structured
+in a series of passes, each of which walks the entire AST. We do not
+need to change this structure to enable incremental
+compilation. Instead, we continue to do every pass as normal, but when
+we come to an item for which we have a pre-existing artifact (for
+example, if we are type-checking a fn that has not changed since the
+last execution), we can simply skip over that fn instead. Similar
+strategies can be used to enable lazy or parallel compilation at later
+times. (Eventually, though, it might be nice to restructure the
+compiler so that it operates in more of a demand driven style, rather
+than a series of sweeping passes.)
+
+When we come to the final LLVM stages, we must
+[separate the functions into distinct "codegen units"](#optimization)
+for the purpose of LLVM code generation. This will build on the
+existing "codegen-units" used for parallel code generation. LLVM may
```

Comment:
I think, it doesn't make much sense to allow the user to specify the number of codegen-units when compiling incrementally. The compiler needs more control over what ends up where and, at least in the beginning, I would expect the `-Ccodgegen-units` option to be ignored (with a note to the user) for incremental builds.